src/python_routing_v01/compute_nhd_routing_SingleSeg.py

```diff
@@ -302,52 +322,256 @@ def compute_mc_reach_up2down(
             velp=velp,
             depthp=depthp,
         )
-        # print(qdc, velc, depthc)
-        # print(qdc_expected, velc_expected, depthc_expected)
-
-        if write_output:
-            write_buffer.append(
-                ",".join(
-                    map(
-                        str,
-                        (
-                            current_segment,
-                            qdp,
-                            depthp,
-                            velp,
-                            qlat,
-                            qup,
-                            quc,
-                            qdc,
-                            depthc,
-                            velc,
-                        ),
-                    )
-                )
-            )
 
         # for next segment qup / quc use the previous flow values
-        current_flow["flow"]["curr"] = qdc
-        current_flow["depth"]["curr"] = depthc
-        current_flow["vel"]["curr"] = velc
+        if ts > 0:
+            qup = flowveldepth[current_segment]["flowval"][-1]  # input for next segment
+        else:
+            qup = 0
 
-        quc = qdc
-        qup = qdp
+        quc = qdc  # input for next segment
+        if assume_short_ts:
+            quc = qup
 
+        flowveldepth[current_segment]["flowval"].append(qdc)
+        flowveldepth[current_segment]["depthval"].append(depthc)
+        flowveldepth[current_segment]["velval"].append(velc)
+        flowveldepth[current_segment]["time"].append(ts * dt)
+
+        next_segment = connections[current_segment]["downstream"]
         if current_segment == reach["reach_tail"]:
-            if verbose:
+            if debuglevel <= -2:
                 print(f"{current_segment} (tail)")
             break
+        if debuglevel <= -3:
+            print(f"{current_segment} --> {next_segment}\n")
+        # current_segment = next_segment
+        # next_segment = connections[current_segment]["downstream"]
+        current_segment = next_segment
+    # end loop initialized the MC vars
+    # end while loop
+
+
+# ### Psuedocode
+# Write Array  to CSV file
+# arguments reach , pathToOutputFile
+# using global connections and flowveldepth.
+def printarray(
+    reach=None, verbose=False, debuglevel=0, pathToOutputFile="../../test/output/text"
+):
+    global connections
```

Comment:
These global statements are only needed if we are intending to mutate them. They are not needed if we only read from them.

Reply:
They also help with the parallel processing -- as long as we run on Linux, which uses copy-on-write, then the global dataset  is never duplicated when we pass it to the various versions of the parallel execution call. That was the deeper idea behind having it that way in the first place.

Reply:
But -- since the globals inside the parallel call are being used in a nested function (to print results for the different networks, I think the functional use actually makes sense. Some simple tests seemed to show that it doesn't mess up the parallel optimization, so I've made this change. We can consider going back fairly easily, if needed. Here's the diff: 

```
diff --git a/src/python_routing_v01/compute_nhd_routing_SingleSeg.py b/src/python_routing_v01/compute_nhd_routing_SingleSeg.py
index f81fc8c..d9eb4b2 100644
--- a/src/python_routing_v01/compute_nhd_routing_SingleSeg.py
+++ b/src/python_routing_v01/compute_nhd_routing_SingleSeg.py
@@ -224,7 +224,9 @@ def compute_network(
     if writeToCSV:
         for x in range(network["maximum_reach_seqorder"], -1, -1):
             for head_segment, reach in ordered_reaches[x]:
-                printarray(
+                writeArraytoCSV(
+                    connections=connections,
+                    flowveldepth=flowveldepth,
                     reach=reach,
                     verbose=verbose,
                     debuglevel=debuglevel,
@@ -233,6 +235,8 @@ def compute_network(

     if writeToNETCDF:
         writeArraytoNC(
+            connections=connections,
+            flowveldepth=flowveldepth,
             network=network,
             nts=nts,
             dt=dt,
@@ -356,11 +360,14 @@ def compute_mc_reach_up2down(
 # Write Array  to CSV file
 # arguments reach , pathToOutputFile
 # using global connections and flowveldepth.
-def printarray(
-    reach=None, verbose=False, debuglevel=0, pathToOutputFile="../../test/output/text"
+def writeArraytoCSV(
+    flowveldepth=None,
+    connections=None,
+    reach=None,
+    verbose=False,
+    debuglevel=0,
+    pathToOutputFile="../../test/output/text",
 ):
-    global connections
-    global flowveldepth

     # define CSV file Header
     header = [["time", "qlat", "q", "v", "d"]]
@@ -401,6 +408,8 @@ def printarray(
 # arguments network, number of timesteps (nts),  timestep in seconds  (dt),  pathToOutputFile
 # using global connections and flowveldepth.
 def writeArraytoNC(
+    flowveldepth=None,
+    connections=None,
     network=None,
     nts=0,
     dt=60,
@@ -408,8 +417,6 @@ def writeArraytoNC(
     debuglevel=0,
     pathToOutputFile="../../test/output/text/",
 ):
-    global connections
-    global flowveldepth
     # create  array variables to copy from python "flowveldepth" which is global
     flowveldepth_data = {
         "segment": [],
```